src/libraries/Microsoft.Win32.Registry/src/Microsoft/Win32/RegistryKey.Windows.cs

```diff
@@ -520,263 +522,235 @@ private unsafe string[] GetValueNamesCore(int values)
             return names.ToArray();
         }
 
-        private object? InternalGetValueCore(string? name, object? defaultValue, bool doNotExpand)
+        private unsafe object? InternalGetValueCore(string? name, object? defaultValue, bool doNotExpand)
         {
-            object? data = defaultValue;
-            int type = 0;
-            int datasize = 0;
-
-            int ret = Interop.Advapi32.RegQueryValueEx(_hkey, name, null, ref type, (byte[]?)null, ref datasize);
+            if (IsPerfDataKey())
+            {
+                return GetPerfDataKeyValue(name);
+            }
 
-            if (ret != 0)
+            // Create an initial stack buffer large enough to satisfy many reg keys.  We need to call RegQueryValueEx
+            // in order to determine the type of the value, and we can avoid further retries if all of the data can be
+            // retrieved in that single call with this buffer.  If we do need to grow, we grow into a pooled array. The
+            // caller is always handed back a copy of this data, either in an array, a string, or a boxed integer.
+            Span<byte> span = stackalloc byte[512];
+            byte[]? pooledArray = null;
+            try
             {
-                if (IsPerfDataKey())
+                // Loop in case we need to try again with a larger buffer size.
+                while (true)
                 {
-                    int size = 65000;
-                    int sizeInput = size;
-
-                    int r;
-                    byte[] blob = new byte[size];
-                    while (Interop.Errors.ERROR_MORE_DATA == (r = Interop.Advapi32.RegQueryValueEx(_hkey, name, null, ref type, blob, ref sizeInput)))
+                    int type = 0;
+                    int result;
+                    int dataLength = span.Length;
+                    fixed (byte* lpData = &MemoryMarshal.GetReference(span))
                     {
-                        if (size == int.MaxValue)
+                        result = Interop.Advapi32.RegQueryValueEx(_hkey, name, null, ref type, lpData, ref dataLength);
+                        if (dataLength < 0)
                         {
-                            // ERROR_MORE_DATA was returned however we cannot increase the buffer size beyond Int32.MaxValue
-                            Win32Error(r, name);
+                            Debug.Fail($"Expected non-negative data size, got {dataLength}.");
```

Comment:
`RegQueryValueEx` returns `uint` (the PInvoke signature has mismatch), so this actually means that the data is more than 2GB. I think it would make more sense to throw ArithmeticException in this case. It is probably hypothetical case - I doubt that registry value can be more than 2GB.